src/query/ts/m3db/encoded_step_iterator_test.go

```diff
@@ -268,3 +310,159 @@ func TestConsolidatedStepIteratorSplitByBlock(t *testing.T) {
+	testConsolidatedStepIteratorSplitByBlock(t, true)
+}
+
+func TestConsolidatedStepIteratorSplitByBlockSequential(t *testing.T) {
+	testConsolidatedStepIteratorSplitByBlock(t, false)
+}
+
+func benchmarkSingleBlock(b *testing.B, withPools bool) {
+	opts := NewOptions().
+		SetLookbackDuration(1 * time.Minute).
+		SetSplitSeriesByBlock(false)
+	require.NoError(b, opts.Validate())
+	if withPools {
+		opts = withPool(b, opts)
+	}
+
+	for i := 0; i < b.N; i++ {
+		for _, tt := range consolidatedStepIteratorTests {
+			b.StopTimer()
+			blocks, _ := generateBlocks(b, tt.stepSize, opts)
+			b.StartTimer()
+			for _, block := range blocks {
+				iters, _ := block.StepIter()
+				for iters.Next() {
+					// no-op
+				}
+
+				require.NoError(b, iters.Err())
+			}
+		}
+	}
+}
+
+func BenchmarkSingleBlockParallel(b *testing.B) {
+	benchmarkSingleBlock(b, true)
+}
+
+func BenchmarkSingleBlockSerialll(b *testing.B) {
+	benchmarkSingleBlock(b, false)
+}
+
+type noopCollector struct{}
+
+func (n noopCollector) AddPoint(dp ts.Datapoint) {}
+
+func benchmarkNextIteration(b *testing.B, iterations int, usePools bool) {
+	ctrl := gomock.NewController(b)
+	defer ctrl.Finish()
+
+	var (
+		seriesCount = 1
+		start       = time.Now()
+		stepSize    = time.Second * 10
+		annotation  = ts.Annotation{}
+		iters       = make([]encoding.SeriesIterator, seriesCount)
+		pIters      = make([]*encoding.MockSeriesIterator, seriesCount)
+		collectors  = make([]consolidators.StepCollector, seriesCount)
+		peeks       = make([]peekValue, seriesCount)
+
+		points = make([]ts.Datapoint, iterations)
+	)
+
+	for i := 0; i < seriesCount; i++ {
+		collectors[i] = noopCollector{}
+		it := encoding.NewMockSeriesIterator(ctrl)
+		it.EXPECT().Err().Return(nil).AnyTimes()
+		iters[i] = it
+		pIters[i] = it
+	}
+
+	for i := 0; i < iterations; i++ {
+		points[i] = ts.Datapoint{Timestamp: start.Add(time.Duration(i) * 5 * time.Second)}
+	}
+
+	duration := stepSize * time.Duration(iterations)
+	it := &encodedStepIterWithCollector{
+		stepTime: start,
+		blockEnd: start.Add(duration),
+		meta: block.Metadata{
+			Bounds: models.Bounds{
+				Start:    start,
+				StepSize: stepSize,
+				Duration: duration,
+			},
+		},
+
+		seriesCollectors: collectors,
+		seriesPeek:       peeks,
+		seriesIters:      iters,
+	}
+
+	if usePools {
+		opts := xsync.NewPooledWorkerPoolOptions()
+		readWorkerPools, err := xsync.NewPooledWorkerPool(1024, opts)
+		require.NoError(b, err)
+		readWorkerPools.Init()
+		it.workerPool = readWorkerPools
+	}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		it.stepTime = start
+		it.finished = false
+		for i := range it.seriesPeek {
+			it.seriesPeek[i] = peekValue{}
+		}
+
+		for _, it := range pIters {
+			for _, p := range points {
+				it.EXPECT().Current().Return(p, xtime.Second, annotation)
+			}
+
+			it.EXPECT().Next().DoAndReturn(func() bool {
+				// simulate load
+				time.Sleep(time.Millisecond)
```

Comment:
Hm instead of mock iterators how about just encoding a bunch into real M3TSZ encoders?  

Not sure if time sleeping is a good proxy for real decoding work since always waiting for a fixed amount of time at each point will have different scheduling characteristics vs actually doing CPU work during a call to Next() on a single iterator.

Reply:
Yeah totally fair, i'll try and do one with no mocks